nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -529,4 +530,369 @@ impl DataStore {
+
+    /// Attempts to lock an instance's record to apply state updates in an
+    /// instance-update saga, if the pprovided `current_gen` matches the
+    /// instance's current `updater_gen`.
+    ///
+    /// This function will attempt to set the `updater_id` field on the record
+    /// corresponding to the provided `authz_instance`  to the provided
+    /// `saga_lock_id`. If the instance's `updater_gen` field is equal to
+    /// `current_gen`, and the instance's `updater_id` field is null, then the
+    /// generation is advanced and the `updater_id` field is set to
+    /// `saga_lock_id`, acquiring the lock for the calling saga. Otherwise, if
+    /// the generation has advanced since `current_gen` was captured, or if
+    /// another saga has locked the instance, the lock is not acquired.
+    ///
+    /// # Notes
+    ///
+    /// This method MUST only be called from the context of a saga! The
+    /// calling saga must ensure that the reverse action for the action that
+    /// acquires the lock must call [`DataStore::instance_updater_unlock`] to
+    /// ensure that the lock is always released if the saga unwinds.
+    ///
+    /// # Arguments
+    ///
+    /// - `authz_instance`: the instance to attempt to lock to lock
+    /// - `current_gen`: the current generation of the instance's `updater_id`
+    /// - `saga_lock_id`: the UUID of the saga that's attempting to lock this
+    ///   instance.
+    ///
+    /// # Returns
+    ///
+    /// - [`Ok`]`((`[Instance`]`, true))` if the lock was acquired.
+    /// - [`Ok`]`((`[Instance`]`, false))` if the lock was not acquired because
+    ///   the instance was already locked by another saga, or because the
+    ///   generation has advanced since `current_gen` was captured.
+    /// - [`Err`]`(`[`Error::ObjectNotFound`]`)` if no instance with the
+    ///   provided ID exists (or the instance record has been deleted).
+    /// - [`Err`]`(other)` if the database query failed for another reason.
+    pub async fn instance_updater_try_lock(
+        &self,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
+        current_gen: Generation,
+        saga_lock_id: &Uuid,
+    ) -> Result<(Instance, bool), Error> {
+        use db::schema::instance::dsl;
+
+        // The generation to advance to.
+        let new_gen = Generation(current_gen.0.next());
+
+        let instance_id = authz_instance.id();
+
+        let locked = diesel::update(dsl::instance)
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::id.eq(instance_id))
+            // If the generation is the same as the captured generation when we
+            // read the instance record to check if it was not locked, we can
+            // lock this instance. Otherwise, the update will fail. This query
+            // is equivalent to an atomic compare-and-swap instruction in a
+            // non-distributed single-process mutex.
+            .filter(dsl::updater_gen.eq(current_gen))
```

Comment:
In other words: There's an assumption that if `updater_gen` is the same, the **value** of `updater_id` is `None`?